04 feb2023

AnShYu opened the pull request:
WeatherStation added

AndShir committed:
RunningHistory v.1

AndShir committed:
RunningHistory v.1

AndShir committed:
RunningHistory v.1

annaalkh commented on src/ru/andrey/weatherstation/Forecast.java:
```diff
@@ -0,0 +1,18 @@
+package ru.andrey.weatherstation;
+
```
123

annaalkh commented on src/ru/andrey/weatherstation/Forecast.java:
```diff
@@ -0,0 +1,18 @@
+import java.time.LocalDate;
+
+public class Forecast {
+    Float temperature;
```
У полей обязательно нужно указать уровень доступа - в данном случае private

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,73 @@
+    Float latitude;
+    Float longitude;
+    private String type;
+    private Boolean normalFunctioning; // могу ли сделать поля не прайват? или же для всех полей, к которым хочу обращаться из других файлов, нужно делать геттер?
```
Если у класса предполагается наличие наследников и им нужен доступ к этому полю, его можно сделать protected, в остальных случаях поля должны быть private

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,73 @@
+    String name;
+    Float latitude;
+    Float longitude;
+    private String type;
```
В чем здесь принципиальная проблема: у тебя тип радара определяется одновременно и классом (который из наследников), и полем "тип" - по сути, один и тот же факт дублируется дважды, что плохо. Нужно оставить либо поле с типом, либо наследование.
С одной стороны, с точки зрения ООП наследование правильнее, в этом и смысл зачем его придумали - чтобы в случаях, когда логика поведения либо набор необходимых полей зависит от подвида, не городить логику на основе полей типа "тип", а просто реализовать ее по-разному в разных наследниках.
С другой стороны - никакой логики пока в данном случае в наследниках не содержится, по факту у тебя три одинаковых класса, отличающихся только названием. А в таком случае, если разницы между ними нет и не предвидится, лучше и проще будет сделать один класс с полем "тип"

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,73 @@
+        this.normalFunctioning = true;
+    }
+
+    public void isMalfunctioning () {
```
Плохое название метода - методы с названием isЧто-то обычно проверяют факт и возвращают boolean-значение (как метод isNormalFunctioning ()), а не меняют данные - здесь лучше будет назвать метод, например, setMulfunctioning. Тем более у тебя в одном классе получается два метода с похожим наименованием, но работающие принципиально по-разному - это совсем плохо

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,73 @@
+            int numberOfDaysInCalculation = 5;
+            Float totalOfRadarReadings = 0.0f;
+            for (int i = 1; i <= numberOfDaysInCalculation; i++) {
+                totalOfRadarReadings = totalOfRadarReadings + mapOfTheRadarReadings.get(date.minusDays(i)).readingValue;
```
Если за какой-то день не содержится данных, у тебя эта строчка красиво упадет
Плюс доступ к полям всегда должен быть через геттеры

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,73 @@
+
+    public Float getAverageReadingForPeriod (LocalDate date) {
+        if (normalFunctioning) {
+            int numberOfDaysInCalculation = 5;
```
5 - это константа, ее стоит вынести в public static final поле

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    public Boolean thereAreReadingsForAllDaysInCalculation  (LocalDate date) {
+            int numberOfDaysInCalculation = 5;
```
Тем более здесь 5 тоже используется

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,73 @@
+
+    public Boolean thereAreReadingsForAllDaysInCalculation  (LocalDate date) {
+            int numberOfDaysInCalculation = 5;
+            Boolean thereAreReadingsForAllDaysInCalculation = true;
```
Всегда стоит использовать примитивы, если только у тебя нет явной причины использовать классы-обертки - в данном случае ее нет

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,73 @@
+            int numberOfDaysInCalculation = 5;
+            Boolean thereAreReadingsForAllDaysInCalculation = true;
+            for (int i = 1; i <= numberOfDaysInCalculation; i++) {
+                if (mapOfTheRadarReadings.containsKey(date.minusDays(i)) == false) {
```
Проверка на false лишняя - достаточно просто if (!mapOfTheRadarReadings.containsKey(date.minusDays(i)))

annaalkh commented on src/ru/andrey/weatherstation/RadarReading.java:
```diff
@@ -0,0 +1,22 @@
+import java.time.LocalDate;
+
+public class RadarReading implements Comparable<RadarReading> {
+    LocalDate date;
```
private

annaalkh commented on src/ru/andrey/weatherstation/UidGenerator.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class UidGenerator {
```
Когда используются классы - генераторы чего-то, обычно принцип их работы следующий: ты где-то при старте программы создаешь экземпляр этого класса, и затем, когда тебе нужно получить следующий uid, вызываешь у него метод  String getNextUid(String prefix), который и вернет тебе uid. А если ты, как сейчас, при каждом создании радара создаешь новый экземпляр UidGenerator, никакой нумерации не получится - она каждый раз будет начинаться с 0

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,104 @@
+
+    public List<RadarReading> getAllReadingsOfTheRadar (String uid) {
+        Radar requestedRadar = mapOfRadars.get(uid);
+        List<RadarReading> arrayOfTheRadarReadings = new ArrayList<>(requestedRadar.mapOfTheRadarReadings.values());
```
У тебя тип переменной - List, а название - array, лучше чтобы такого не было

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,104 @@
+        int numberOfTemperatureRadars = 0;
+        int numberOfHumidityRadars = 0;
+        int numberOfWindSpeedRadars = 0;
+        int numberOfTemperatureRadarsWithReadingsForAllDaysInCalculation = 0;
```
В java, конечно, любят длинные имена - но настолько - не стоит:)

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,104 @@
+        int numberOfHumidityRadarsWithReadingsForAllDaysInCalculation = 0;
+        int numberOfWindSpeedRadarsWithReadingsForAllDaysInCalculation = 0;
+        for (String uid: mapOfRadars.keySet()) {
+            if (mapOfRadars.get(uid).getType().equals("Temperature")) {
```
Постарайся сделать так, чтобы вот этого if-а не было - это не очень хороший паттерн. Как минимум потому, что у тебя практически один и тот же код дублируется три раза

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,104 @@
+        mapOfRadars.get(uid).isNotMalfunctioning();
+    }
+
+    public Map<String, Radar> getMapOfAllMalfunctioningRadars () {
```
По условию, тебе нужно вернуть список

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,73 @@
+        normalFunctioning = true;
+    }
+
+    public void addRadarReading (LocalDate date, RadarReading radarReading) {
```
Для меня немного сомнительная идея - хранить данные о показаниях радара в классе самого радара. По логике, все данные хранятся в метеостанции, а задача радара - зафиксировать данные и отправить их на метеостанцию. Можно и так, в принципе - каких-то существенных неувязок не вижу, но все же подумай еще раз:)

AndShir committed:
Changed in accordance with Anya's comments

AndShir committed:
Added exceptions

AndShir committed:
Added class Main

AndShir committed:
Added class Main

AndShir committed:
Added class Main

AndShir committed:
Fixed problem with forecast calculation

AndShir committed:
Sieve of Eratosthenes

AndShir committed:
1. Made a map for each radar type; 2. reduced number of calls to maps; 3. changed wrapper classes to primitive data types; 4. Added WrongRadarUIDException

AndShir committed:
Added specificParameterForecast and changed method for forecast calculation

AndShir committed:
(1) UIDGenerator changed to singleton;
(2) Added reading of RadarReadings from file;
(3) Added saving of forecasts to file;
(4) Added saving radars to file before stopping the program and reusing them when newly started

AndShir committed:
(1) UIDGenerator changed to singleton;
(2) Added reading of RadarReadings from file;
(3) Added saving of forecasts to file;
(4) Added saving radars to file before stopping the program and reusing them when newly started

AndShir committed:
(1) UIDGenerator changed to singleton;
(2) Added reading of RadarReadings from file;
(3) Added saving of forecasts to file;
(4) Added saving radars to file before stopping the program and reusing them when newly started

AndShir committed:
(1) UIDGenerator changed to singleton;
(2) Added reading of RadarReadings from file;
(3) Added saving of forecasts to file;
(4) Added saving radars to file before stopping the program and reusing them when newly started

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,106 @@
+    private float longitude;
+    private String type;
+
+    private static final int DAYS_IN_ACCOUNT_FOR_FORECAST = 5;
```
Константы лучше размещать отдельно от остальных полей, лучше всего в начале класса

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,106 @@
+
+
+    public Float getAverageReadingForPeriod (LocalDate date) {
+            Float totalOfRadarReadings = 0.0f;
```
Почему не float?

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,106 @@
+                    daysWithReadings++;
+                }
+            }
+            Float averageReading = totalOfRadarReadings / daysWithReadings;
```
И здесь float. И в сигнатуре метода тоже

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,106 @@
+        return uid;
+    }
+
+    public Map<LocalDate, RadarReading> getMapOfTheRadarReadings() {
```
В идеале такого метода быть не должно - хэш мапа это деталь внутренней реализации класса

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,106 @@
+    private String name;
+    private float latitude;
+    private float longitude;
+    private String type;
```
Было бы правильнее сделать тип enum-ом

annaalkh commented on src/ru/andrey/weatherstation/Radar.java:
```diff
@@ -0,0 +1,106 @@
+
+    private static final int DAYS_IN_ACCOUNT_FOR_FORECAST = 5;
+    private boolean normalFunctioning;
+    private Map<LocalDate, RadarReading> mapOfTheRadarReadings = new HashMap<>();
```
Когда мы экспортируем в файл список радаров, список их прочитанных значений не должен сохраняться - по условию требуется сериализовать и потом восстанавливать только данные о самих радарах, а импорт сохраненных значений - отдельная задача, которую пользователь может и не делать. Сейчас получается, что они пересекаются, поэтому могут возникнуть конфликты

annaalkh commented on src/ru/andrey/weatherstation/RadarReading.java:
```diff
@@ -0,0 +1,58 @@
+    private static final long serialVersionUID = 1L;
+    private LocalDate date;
+    private float readingValue;
+    String uid;
```
А почему uid не private?

annaalkh commented on src/ru/andrey/weatherstation/RadarReadingsFileManager.java:
```diff
@@ -0,0 +1,52 @@
+
+    //дата, uid радара, тип параметра, значение параметра - презюмирую, что каждая запись начинается с новой строчки
+    public static List<RadarReading> makeListOfRadarReadings(File file) {
+        List<String> listOfCSVs = readRadarReadingsFromFile(file);
```
Не совсем корректное имя переменной - csv это формат файла целиком, а не отдельной строки, применять его к отдельной строке выглядит немного странно

annaalkh commented on src/ru/andrey/weatherstation/RadarReadingsFileManager.java:
```diff
@@ -0,0 +1,52 @@
+    public static List<RadarReading> makeListOfRadarReadings(File file) {
+        List<String> listOfCSVs = readRadarReadingsFromFile(file);
+        List<RadarReading> listOfRadarReadings = new ArrayList<>();
+            for (String string : listOfCSVs) {
```
Форматирование

annaalkh commented on src/ru/andrey/weatherstation/RadarReadingsFileManager.java:
```diff
@@ -0,0 +1,52 @@
+        List<String> listOfCSVs = readRadarReadingsFromFile(file);
+        List<RadarReading> listOfRadarReadings = new ArrayList<>();
+            for (String string : listOfCSVs) {
+                String[] parts = string.split("\\,");
```
Не совсем уверена - точно ли здесь нужен \\?

annaalkh commented on src/ru/andrey/weatherstation/RadarReadingsFileManager.java:
```diff
@@ -0,0 +1,52 @@
+            }
+            return listOfRadarReadings;
+
+//            int commaIndex1 = string.indexOf(",");
```
Закомментированный код

annaalkh commented on src/ru/andrey/weatherstation/RadarReadingsFileManager.java:
```diff
@@ -0,0 +1,52 @@
+            while (readValue != null) {
+                listOfReadValues.add(readValue);
+                readValue = br.readLine();
+            }
```
А где возвращение прочитанных значений?

annaalkh commented on src/ru/andrey/weatherstation/RadarReadingsFileManager.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+
+    private static List<String> readRadarReadingsFromFile (File file) {
```
Лишний пробел перед списком аргументов

annaalkh commented on src/ru/andrey/weatherstation/RadarReadingsFileManager.java:
```diff
@@ -0,0 +1,52 @@
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
```
Из методов, которые возвращают коллекции, никогда не стоит возвращать null - лучше пустую коллекцию, например Collections.emptyList(). Так будет гораздо меньше проблем с NPE

annaalkh commented on src/ru/andrey/weatherstation/RadarReadingsFileManager.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class RadarReadingsFileManager {
```
Имя класса не совсем точное - Manager значит, что класс чем-то управляет, у него есть состояние - а в данном случае это просто набор утилитарных методов

annaalkh commented on src/ru/andrey/weatherstation/RadarsFileManager.java:
```diff
@@ -0,0 +1,29 @@
+import java.io.*;
+import java.util.List;
+
+public class RadarsFileManager {
```
То же самое про имя класса

annaalkh commented on src/ru/andrey/weatherstation/RadarsFileManager.java:
```diff
@@ -0,0 +1,29 @@
+        } catch (ClassNotFoundException e) {
+            e.printStackTrace();
+        }
+        return null;
```
Тоже null возвращать не стоит

annaalkh commented on src/ru/andrey/weatherstation/UIDGeneratorSingleton.java:
```diff
@@ -0,0 +1,24 @@
+        return INSTANCE;
+    }
+
+    private int counter = 1;
```
Поле лучше перенести в начало, здесь оно теряется немного

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+
+public class WeatherStation {
+
+    private Map<String, Radar> mapOfTemperatureRadars = new HashMap<>();
```
Еще есть момент - поля, которые не планируется что будут изменяться (а эти мапы изменяться не будут, только их содержимое) - по правилам хорошего кода рекомендуется делать final

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+        this.uidGenerator = UIDGeneratorSingleton.getInstance();
+    }
+
+    public Map<String, Radar> getMapOfTemperatureRadars() {
```
Этих методов тоже лучше бы не было

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    public void addRadar (Radar radar) {
+        uidGenerator.increaseUIDCounter(); // это норм подход? Или лучше получать counter через getter
```
Так работать не будет (ну, или будет, но ненадежно). В файле uid может быть любым, то, что мы увеличим счетчик на один, не дает никаких гарантий, что в какой-то момент мы не получим далее дублирующийся uid. Надо сделать так, чтобы после чтения всех радаров из файла счетчик в UIDGenerator соответствовал максимальному из них. Лучше всего для этого сделать в UIDGenerator метод с именем updateUid, который будет принимать на вход uid из файла и обновлять свой внутренний счетчик в соответствии с ним

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+    public void addRadar (Radar radar) {
+        uidGenerator.increaseUIDCounter(); // это норм подход? Или лучше получать counter через getter
+        switch (radar.getType()) {
+            case "Temperature":
```
enum бы...

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+
+    public void addRadarReading(List<RadarReading> list) {
+        for (RadarReading reading : list) {
+            if (this.mapOfTemperatureRadars.containsKey(reading.getUid())) {
```
this лишний

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+    public List<RadarReading> getAllReadingsOfTheRadar (String uid) throws RadarMalfunctionException, WrongRadarUIDException {
+        Radar radar = getRadar(uid);
+        if (!radar.isNormalFunctioning()) throw new RadarMalfunctionException("Radar is malfunctioning");
+        List<RadarReading> listOfTheRadarReadings = new ArrayList<>(radar.getMapOfTheRadarReadings().values());
```
Было бы лучше, если бы у радара был метод, сразу возвращающий список значений (можно даже отсортированный)

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+
+        Forecast forecast = new Forecast (averageTemperature, averageHumidity, averageWindSpeed, isPrecise);
+
+        try (Writer writer  = new FileWriter(file, true)) {
```
Плохой вариант - смешивать логику вычисления прогноза и его сохранение в файл, тем более что название метода ничего про сохранение не говорит - такой себе сюрприз для пользователя получится. Сохранение в файл нужно вынести в отдельный метод (а еще лучше - в отдельный класс) и вызывать отдельно

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+        float averageHumidity = humidityForecast.getParameterForecast();
+        float averageWindSpeed = windSpeedForecast.getParameterForecast();
+
+        Forecast forecast = new Forecast (averageTemperature, averageHumidity, averageWindSpeed, isPrecise);
```
Было бы красивее, если бы ты передавал в конструктор Forecast все четыре SpecificParameterForecast, а он бы уже сам разбирался, как из них получить конкретные значения, и там же проставлял бы isPrecise

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+                listOfMalfunctioningRadars.add(radar);
+            }
+        }
+        for (String uid: mapOfHumidityRadars.keySet()) {
```
Код дублируется - лучше было бы вынести повторяющийся код в отдельный метод

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+
+    private Radar getRadar(String uid) throws WrongRadarUIDException {
+        Radar radar;
+        if (mapOfTemperatureRadars.containsKey(uid)) {
```
И опять ты проходишь вручную по всем трем хэш-мапам. Можно было бы, например, создать список всех хэшмап и проходить по нему

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+        return mapOfHumidityRadars;
+    }
+
+    public void addRadar (String uidPrefix, String name, float latitude, float longitude, String type) throws WrongRadarTypeException {
```
Лишние пробелы после имени метода по всему файлу

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+        if (numberOfParameterRadars <= 1) {
+            isPrecise = false;
+        }
+        if (numberOfParameterRadarsWithEnoughReadings == 0) {
```
Почему у тебя два одинаковых условия по смыслу, но в первому проверяется <=1, а во втором ==0? И какой еще вариант есть <=1 кроме как 0? Плюс здесь было бы проще объединить оба условия в одно

annaalkh commented on src/ru/andrey/weatherstation/WeatherStation.java:
```diff
@@ -0,0 +1,218 @@
+
+    public List<Radar> getListOfAllRadars() {
+        List <Radar> listOfRadars = new ArrayList<>();
+        for (Radar radar: this.getMapOfTemperatureRadars().values()) {
```
И третий раз проходим по всем мапам)

annaalkh commented on src/ru/andrey/weatherstation/Main.java:
```diff
@@ -0,0 +1,182 @@
+            main.readAllSavedRadars(main.radarsFile);
+        }
+        if (main.readingsFile.exists()) {
+            main.readAllSavedRadarReadings(main.readingsFile);
```
С точки зрения юзабельности, было бы хорошо вывести пользователю сообщение, что что-то прочитано или не прочитано

annaalkh commented on src/ru/andrey/weatherstation/Main.java:
```diff
@@ -0,0 +1,182 @@
+        } catch (RadarMalfunctionException e) {
+            e.printStackTrace();
+        } catch (WrongRadarUIDException e) {
+            e.printStackTrace();
```
Нехорошо, что в некоторых местах информация об ошибке выводится с помощью System.out.println, а в некоторых просто выкидывается стек трейс - лучше везде использовать System.out.println

annaalkh commented on src/ru/andrey/weatherstation/Main.java:
```diff
@@ -0,0 +1,182 @@
+        try {
+            weatherStation.addRadarReading(uid, date, radarReading);
+        } catch (RadarMalfunctionException e) {
+            e.printStackTrace();
```
А почему в этом случае мы не продолжаем работу? И в следующих методах то же самое

AndShir committed:
Реализация алгоритма шифрования RSA

AndShir committed:
Внес исправления по части из комментариев Ани

AndShir committed:
Внес все изменения по комментариям Ани

AndShir committed:
Поиск строк со словом "Рассвет"

AndShir committed:
Поиск строк со словом "Рассвет"

AndShir committed:
Парсинг тем писем из файла

AndShir committed:
Исправил небольшой баг - не читалась последняя строка из файла. Добавил list.add(line) в строке 116
